src/Abp.Web/Web/AbpModuleAssemblyFindler.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Abp.Collections.Extensions;
+using Abp.Modules;
+using Abp.Reflection;
+
+namespace Abp.Web
+{
+    /// <summary>
+    /// This class is used to get assemblies, which contain at least one AbpModule-derived class.
+    /// </summary>
+    public class AbpModuleAssemblyFindler : IAssemblyFinder
+    {
+        private List<Assembly> _assemblies;
+
+        private readonly object _syncLock = new object();
+
+        /// <summary>
+        /// This return assemblies, which contain at least one AbpModule-derived class
+        /// </summary>
+        /// <returns>List of assemblies</returns>
+        public List<Assembly> GetAllAssemblies()
+        {
+            if (_assemblies == null)
+            {
+                lock (_syncLock)
+                {
+                    if (_assemblies == null)
+                    {
+                        _assemblies = GetAbpModuleAssembliesRecursive();
+                    }
+                }
+            }
+
+            return _assemblies;
+        }
+
+        /// <summary>
+        /// Retrieves all of the AbpModules and their dependencies
+        /// </summary>
+        /// <returns></returns>
+        private List<Assembly> GetAbpModuleAssembliesRecursive()
+        {
+            var currentAssembly = Assembly.GetCallingAssembly();
```

Comment:
Okay, let's stick the idea of direct  top level module declaration. Should we just pass in into AbpBoostrapper like a parameter by calling `typeof(ProjectNameWebModule)`? If so, how are we going to register ITypeFinder so it will get an access to this Type?